app/routes/createObservation.tsx

```diff
@@ -0,0 +1,60 @@
+import { json } from "@remix-run/node";
+import { createObservation } from "~/utils/db.server";
+import { createObservation as uploadToS3, genKey } from "~/utils/s3.server";
+
+type Location = { x: number; y: number };
+
+export const action = async ({ request }: { request: Request }) => {
+  if (request.method === "POST") {
+    const formData = await request.json();
+
+    const {
+      title,
+      description,
+      location,
+      picture,
+    }: {
+      title: string;
+      description: string;
+      location: Location;
+      picture: { data: string; ext: string } | null;
+    } = formData;
+
+    if (!title || !description || !location) {
+      return json({ error: "Missing required fields" }, { status: 400 });
+    }
+
+    let pictureKey: string | null = null;
+
+    if (picture && picture.data && picture.ext) {
+      try {
+        const generatedKey = genKey(picture.ext);
+
+        const imageData = Uint8Array.from(atob(picture.data), (char) =>
+          char.charCodeAt(0),
+        );
+
+        pictureKey = await uploadToS3(imageData, generatedKey);
```

Comment:
Maybe the documentation wasn't clear about this; don't generate a key here, the key is automatically generated in this function. You only need to pass the extension as second argument.

Reply:
I just thought it would be nice for the `genKey` function to be public, but in most cases you shouldn't be using it.

Reply:
Should be corrected.